src/libs/opts/opts.c

```diff
@@ -0,0 +1,951 @@
+ * instead a behaviour similar that used for PATH is adopted. On Windows the variable will be split at
+ * each ';' character. On all other systems ':' is used as a separator.
+ *
+ * In case multiple versions (short, long, env-var) of an option are found, the order of precedence is:
+ * <ul>
+ * 	<li> Short options always win. </li>
+ * 	<li> Long options are used if no short option is present. </li>
+ * 	<li> If neither a long nor short option is found, environment variables are considered. </li>
+ * </ul>
+ *
+ * Lastly, all remaining elements of @p argv will be collected into an array. You can access this array
+ * by specifying `args = remaining` on a key with basename '#'. The array will be copied into this key. As is
+ * the case with getopt(3) processing of options will stop if '--' is encountered in @p argv.
+ *
+ * NOTE: Both options and environment variables can only be specified on a single key. This is because
+ * otherwise e.g. the help message ('opt/help', 'env/help') may be ambiguous. If you need to have the
+ * value of one option/environment variable in multiple keys, you may use fallbacks.
+ *
+ * NOTE: Per default option processing DOES NOT stop, when a non-option string is encountered in @p argv.
+ * If you want processing to stop, set the metadata `posixly = 1` on @p errorKey.
+ *
+ * @param ks	The KeySet containing the specification for the options.
+ * @param argc	The number of strings in argv.
+ * @param argv	The arguments to be processed.
+ * @param envp	A list of environment variables. This needs to be a null-terminated list of
+ * 		strings of the format 'KEY=VALUE'.
+ * @param errorKey A key to store an error in, if one occurs.
+ *
+ * @retval 0	on success
+ * @retval -1	on error, the error will be added to @p errorKey
+ * @retval 1	if help option was found
+ */
+int elektraGetOpts (KeySet * ks, int argc, const char ** argv, const char ** envp, Key * errorKey)
+{
+	KeySet * keysWithOpts = ksNew (0, KS_END);
+	KeySet * usedEnvVars = ksNew (0, KS_END);
+	KeySet * optionsSpec = ksNew (
+		2, keyNew ("/short/h", KEY_META, "hasarg", "none", KEY_META, "kind", "single", KEY_META, "flagvalue", "1", KEY_END),
+		keyNew ("/long/help", KEY_META, "hasarg", "none", KEY_META, "kind", "single", KEY_META, "flagvalue", "1", KEY_END), KS_END);
+
+	cursor_t initial = ksGetCursor (ks);
+
+	ksRewind (ks);
+	Key * cur;
+	while ((cur = ksNext (ks)) != NULL)
+	{
+		if (keyGetNamespace (cur) != KEY_NS_SPEC)
+		{
+			continue;
+		}
+
+		const char * kind = "single";
+		if (strcmp (keyBaseName (cur), "#") == 0)
+		{
+			kind = "array";
+		}
+
+		Key * key = NULL;
+		KeySet * opts = ksMetaGetSingleOrArray (cur, "opt");
+		if (opts != NULL)
+		{
+			ksRewind (opts);
+			ksNext (opts); // skip count TODO: bug?
+			Key * k;
+			while ((k = ksNext (opts)) != NULL)
+			{
+				char metaBuffer[ELEKTRA_MAX_ARRAY_SIZE + 15]; // 14 = opt//flagvalue (1 extra to circumvent warning)
```

Comment:
you can also write sizeof("opt/flagvalue")? (+comment has double slash between opt and flagvalue)

Reply:
The double slash is intentional. We need one slash before the array index and one after.